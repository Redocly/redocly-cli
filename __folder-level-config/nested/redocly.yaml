rules:
  info-contact: error
# extends:
#   - ../redocly.yaml
