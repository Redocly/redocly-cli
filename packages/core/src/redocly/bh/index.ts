import * as client from './api-client';
import * as domains from './domains';
import * as apiKeys from './api-keys';

export namespace BlueHarvest {
  export const ApiClient = client.ApiClient;
  export const getDomain = domains.getDomain;
  export const getApiKeys = apiKeys.getApiKeys;
}
