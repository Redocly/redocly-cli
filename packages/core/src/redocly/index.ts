import { existsSync, readFileSync, writeFileSync, unlinkSync } from 'fs';
import { resolve } from 'path';
import { homedir } from 'os';
import { yellow, red, green, gray } from 'colorette';
import { query } from './query';

const TOKEN_FILENAME = '.redocly-config.json';

export class RedoclyClient {
  private accessToken: string | undefined;

  constructor() {
    this.loadToken();
  }

  hasToken(): boolean {
    return !!this.accessToken;
  }

  loadToken(): void {
    if (process.env.REDOCLY_AUTHORIZATION) {
      this.accessToken = process.env.REDOCLY_AUTHORIZATION;
      return;
    }

    const credentialsPath = resolve(homedir(), TOKEN_FILENAME);
    if (existsSync(credentialsPath)) {
      const credentials = JSON.parse(readFileSync(credentialsPath, 'utf-8'));
      this.accessToken = credentials && credentials.token;
    }
  }

  async isAuthorizedWithRedocly(): Promise<boolean> {
    return this.hasToken() && !!(await this.getAuthorizationHeader());
  }

  async verifyToken(accessToken: string, verbose: boolean = false): Promise<boolean> {
    if (!accessToken) return false;
    const authDetails = await RedoclyClient.authorize(accessToken, { verbose });
    if (!authDetails) return false;
    return true;
  }

  async getAuthorizationHeader(): Promise<string | undefined> {
    // print this only if there is token but invalid
    if (this.accessToken && !(await this.verifyToken(this.accessToken))) {
      process.stderr.write(
        `${yellow(
          'Warning:',
        )} invalid Redocly API key. Use "npx @redocly/openapi-cli login" to provide your API key\n`,
      );
      return undefined;
    }
    return this.accessToken;
  }

  async login(accessToken: string, verbose: boolean = false) {
    const credentialsPath = resolve(homedir(), TOKEN_FILENAME);
    process.stdout.write(gray('\n  Logging in...\n'));

    const authorized = await this.verifyToken(accessToken, verbose);

    if (!authorized) {
      process.stdout.write(
        red('Authorization failed. Please check if you entered a valid API key.\n'),
      );
      process.exit(1);
    }

    this.accessToken = accessToken;
    const credentials = {
      token: accessToken,
    };

    writeFileSync(credentialsPath, JSON.stringify(credentials, null, 2));
    process.stdout.write(green('  Authorization confirmed. ‚úÖ\n\n'));
  }

  logout(): void {
    const credentialsPath = resolve(homedir(), TOKEN_FILENAME);
    if (existsSync(credentialsPath)) {
      unlinkSync(credentialsPath);
    }
    process.stdout.write('Logged out from the Redocly account. ‚úã\n');
  }

  async query(queryString: string, parameters = {}, headers = {}) {
    return query(queryString, parameters, {
      Authorization: this.accessToken,
      ...headers,
    });
  }

  static async authorize(accessToken: string, options: { queryName?: string; verbose?: boolean }) {
    const { queryName = '', verbose = false } = options;
    try {
      const queryStr = `query ${queryName}{ viewer { id } }`;

      return await query(queryStr, {}, { Authorization: accessToken });
    } catch (e) {
      if (verbose) console.log(e);
      return null;
    }
  }

  async updateDependencies(dependencies: string[] | undefined): Promise<void> {
    const definitionId = process.env.DEFINITION;
    const versionId = process.env.DEFINITION;
    const branchId = process.env.BRANCH;

    if (!definitionId || !versionId || !branchId) return;

    await this.query(
      `
    mutation UpdateBranchDependenciesFromURLs(
      $urls: [String!]!
      $definitionId: Int!
      $versionId: Int!
      $branchId: Int!
    ) {
      updateBranchDependenciesFromURLs(
        definitionId: $definitionId
        versionId: $versionId
        branchId: $branchId
        urls: $urls
      ) {
        branchName
      }
    }
    `,
      {
        urls: dependencies || [],
        definitionId: parseInt(definitionId, 10),
        versionId: parseInt(versionId, 10),
        branchId: parseInt(branchId, 10),
      },
    );
  }

  updateDefinitionVersion(definitionId: number, versionId: number, updatePatch: object): Promise<void> {
    return this.query(`
      mutation UpdateDefinitionVersion($definitionId: Int!, $versionId: Int!, $updatePatch: DefinitionVersionPatch!) {
        updateDefinitionVersionByDefinitionIdAndId(input: {definitionId: $definitionId, id: $versionId, patch: $updatePatch}) {
          definitionVersion {
            ...VersionDetails
            __typename
          }
          __typename
        }
      }

      fragment VersionDetails on DefinitionVersion {
        id
        nodeId
        uuid
        definitionId
        name
        description
        sourceType
        source
        registryAccess
        __typename
      }
    `,
      {
        definitionId,
        versionId,
        updatePatch,
      },
    );
  }

  getOrganizationId(organizationId: string) {
    return this.query(`
      query ($organizationId: String!) {
        organizationById(id: $organizationId) {
          id
        }
      }
    `, {
      organizationId
    });
  }

  getDefinitionByName(name: string, organizationId: string) {
    return this.query(`
      query ($name: String!, $organizationId: String!) {
        definition: definitionByOrganizationIdAndName(name: $name, organizationId: $organizationId) {
          id
        }
      }
    `, {
      name,
      organizationId
    });
  }

  createDefinition(organizationId: string, name: string) {
    return this.query(`
      mutation CreateDefinition($organizationId: String!, $name: String!) {
        def: createDefinition(input: {organizationId: $organizationId, name: $name }) {
          definition {
            id
            nodeId
            name
          }
        }
      }
    `, {
      organizationId,
      name
    })
  }

  createDefinitionVersion(definitionId: string, name: string, sourceType: string, source: any) {
    return this.query(`
      mutation CreateVersion($definitionId: Int!, $name: String!, $sourceType: DvSourceType!, $source: JSON) {
        createDefinitionVersion(input: {definitionId: $definitionId, name: $name, sourceType: $sourceType, source: $source }) {
          definitionVersion {
            id
          }
        }
      }
    `, {
      definitionId,
      name,
      sourceType,
      source
    });
  }

  getSignedUrl(organizationId: string, filesHash: string, fileName: string) {
    return this.query(`
      query ($organizationId: String!, $filesHash: String!, $fileName: String!) {
        signFileUploadCLI(organizationId: $organizationId, filesHash: $filesHash, fileName: $fileName) {
          signedFileUrl
          uploadedFilePath
        }
      }
    `, {
      organizationId,
      filesHash,
      fileName
    })
  }

  getDefinitionVersion(organizationId: string, definitionName: string, versionName: string) {
    return this.query(`
      query ($organizationId: String!, $definitionName: String!, $versionName: String!) {
        version: definitionVersionByOrganizationDefinitionAndName(organizationId: $organizationId, definitionName: $definitionName, versionName: $versionName) {
          id
          definitionId
          defaultBranch {
            name
          }
        }
      }
    `, {
      organizationId,
      definitionName,
      versionName
    });
  }

  static isRegistryURL(link: string): boolean {
    const domain = process.env.REDOCLY_DOMAIN || 'redoc.ly';
    if (!link.startsWith(`https://api.${domain}/registry/`)) return false;
    const registryPath = link.replace(`https://api.${domain}/registry/`, '');

    const pathParts = registryPath.split('/');

    // we can be sure, that there is job UUID present
    // (org, definition, version, bundle, branch, job, "openapi.yaml" ü§¶‚Äç‚ôÇÔ∏è)
    // so skip this link.
    // FIXME
    if (pathParts.length === 7) return false;

    return true;
  }
}
