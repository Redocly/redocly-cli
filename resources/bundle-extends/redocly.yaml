# plugins:
#   - 'plugins/my-rules-1.js'
rules:
  my-rules-1/opid-not-test: warn
  
# extends:
#   # - minimal
#   - ./gold.yaml

scorecard:
  levels:
    # - name: Baseline
    #   extends:
    #     - minimal

    # - name: Silver
    #   extends:
    #     - recommended
    #   rules:
    #     my-rules-1/opid-not-test: warn

    - name: Gold
      extends:
        - ./gold.yaml