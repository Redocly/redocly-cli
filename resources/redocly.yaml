plugins: 
  - ./plugin.js
rules:
  plugin1/operation-id-not-test: error
  plugin2/operation-id-not-test: error