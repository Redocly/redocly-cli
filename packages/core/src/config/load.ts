import * as fs from 'fs';
import * as path from 'path';
import { RedoclyClient } from '../redocly';
import { isEmptyObject, loadYaml } from '../utils';
import { Config, DOMAINS } from './config';
import { transformConfig } from './utils';
import { resolveConfig } from './config-resolvers';

import type { RawConfig, Region } from './types';

export async function loadConfig(
  configPath: string | undefined = findConfig(),
  customExtends?: string[],
): Promise<Config> {
  const rawConfig = await getConfig(configPath);
  if (customExtends !== undefined) {
    rawConfig.lint = rawConfig.lint || {};
    rawConfig.lint.extends = customExtends;
  } else if (isEmptyObject(rawConfig)) {
    // TODO: check if we can add recommended here. add message here?
    // rawConfig.lint = { extends: ['recommended'], recommendedFallback: true };
  }

  const redoclyClient = new RedoclyClient();
  const tokens = await redoclyClient.getTokens();

  if (tokens.length) {
    if (!rawConfig.resolve) rawConfig.resolve = {};
    if (!rawConfig.resolve.http) rawConfig.resolve.http = {};
    rawConfig.resolve.http.headers = [...(rawConfig.resolve.http.headers ?? [])];

    for (const item of tokens) {
      const domain = DOMAINS[item.region as Region];
      rawConfig.resolve.http.headers.push(
        {
          matches: `https://api.${domain}/registry/**`,
          name: 'Authorization',
          envVariable: undefined,
          value: item.token,
        },
        //support redocly.com domain for future compatibility
        ...(item.region === 'us'
          ? [
              {
                matches: `https://api.redoc.ly/registry/**`,
                name: 'Authorization',
                envVariable: undefined,
                value: item.token,
              },
            ]
          : []),
      );
    }
  }

  return resolveConfig(rawConfig, configPath);
}

export const CONFIG_FILE_NAMES = ['redocly.yaml', 'redocly.yml', '.redocly.yaml', '.redocly.yml'];

export function findConfig(dir?: string): string | undefined {
  if (!fs.hasOwnProperty('existsSync')) return;
  const existingConfigFiles = CONFIG_FILE_NAMES.map((name) =>
    dir ? path.resolve(dir, name) : name,
  ).filter(fs.existsSync);
  if (existingConfigFiles.length > 1) {
    throw new Error(`
      Multiple configuration files are not allowed. 
      Found the following files: ${existingConfigFiles.join(', ')}. 
      Please use 'redocly.yaml' instead.
    `);
  }
  return existingConfigFiles[0];
}

export async function getConfig(configPath: string | undefined = findConfig()) {
  if (!configPath) return {};
  try {
    const rawConfig = ((await loadYaml(configPath)) || {}) as RawConfig;
    return transformConfig(rawConfig);
  } catch (e) {
    throw new Error(`Error parsing config file at '${configPath}': ${e.message}`);
  }
}
