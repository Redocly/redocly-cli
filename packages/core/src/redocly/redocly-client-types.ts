import { Region } from '../config/config';

export interface RegionalToken {
  region: Region;
  token: string;
}

export interface RegionalTokenWithValidity extends RegionalToken {
  valid: boolean;
}
