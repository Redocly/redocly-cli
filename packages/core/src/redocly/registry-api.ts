import fetch, { RequestInit, HeadersInit } from 'node-fetch';
import { RegistryApiTypes } from './registry-api-types';
import { DEFAULT_REGION, DOMAINS } from '../config/config';
import { isNotEmptyObject } from '../utils';
const version = require('../../package.json').version;

import type { AccessTokens, Region } from '../config/types';

export class RegistryApi {
  constructor(private accessTokens: AccessTokens, private region: Region) {}

  get accessToken() {
    return isNotEmptyObject(this.accessTokens) && this.accessTokens[this.region];
  }

  getBaseUrl(region: Region = DEFAULT_REGION) {
    return `https://api.${DOMAINS[region]}/registry`;
  }

  setAccessTokens(accessTokens: AccessTokens) {
    this.accessTokens = accessTokens;
    return this;
  }

  private async request(path = '', options: RequestInit = {}, region?: Region) {
    const headers = Object.assign({}, options.headers || {}, { 'x-redocly-cli-version': version });

    if (!headers.hasOwnProperty('authorization')) {
      throw new Error('Unauthorized');
    }

    const response = await fetch(
      `${this.getBaseUrl(region)}${path}`,
      Object.assign({}, options, { headers }),
    );

    if (response.status === 401) {
      throw new Error('Unauthorized');
    }

    if (response.status === 404) {
      const body: RegistryApiTypes.NotFoundProblemResponse = await response.json();
      throw new Error(body.code);
    }

    return response;
  }

  async authStatus(
    accessToken: string,
    region: Region,
    verbose = false,
  ): Promise<{ viewerId: string; organizations: string[] }> {
    try {
      const response = await this.request('', { headers: { authorization: accessToken } }, region);

      return await response.json();
    } catch (error) {
      if (verbose) {
        console.log(error);
      }

      throw error;
    }
  }

  async prepareFileUpload({
    organizationId,
    name,
    version,
    filesHash,
    filename,
    isUpsert,
  }: RegistryApiTypes.PrepareFileuploadParams): Promise<RegistryApiTypes.PrepareFileuploadOKResponse> {
    const response = await this.request(
      `/${organizationId}/${name}/${version}/prepare-file-upload`,
      {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          authorization: this.accessToken,
        } as HeadersInit,
        body: JSON.stringify({
          filesHash,
          filename,
          isUpsert,
        }),
      },
      this.region,
    );

    if (response.ok) {
      return response.json();
    }

    throw new Error('Could not prepare file upload');
  }

  async pushApi({
    organizationId,
    name,
    version,
    rootFilePath,
    filePaths,
    branch,
    isUpsert,
    isPublic,
    batchId,
    batchSize
  }: RegistryApiTypes.PushApiParams) {
    const response = await this.request(
      `/${organizationId}/${name}/${version}`,
      {
        method: 'PUT',
        headers: {
          'content-type': 'application/json',
          authorization: this.accessToken,
        } as HeadersInit,
        body: JSON.stringify({
          rootFilePath,
          filePaths,
          branch,
          isUpsert,
          isPublic,
          batchId,
          batchSize
        }),
      },
      this.region,
    );

    if (response.ok) {
      return;
    }

    throw new Error('Could not push api');
  }
}
