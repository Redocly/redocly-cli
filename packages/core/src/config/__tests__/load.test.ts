import { loadConfig } from '../load';
import { RedoclyClient } from '../../redocly';

describe('loadConfig', () => {
  it('should resolve config http header by US region', async () => {
    jest.spyOn(RedoclyClient.prototype, 'getTokens').mockImplementation(
      () => Promise.resolve([{ region: 'us', token: "accessToken", valid: true }])
    );
    const config = await loadConfig();
    expect(config.resolve.http.headers).toStrictEqual([{
      "matches": 'https://api.redoc.ly/registry/**',
      "name": "Authorization",
      "envVariable": undefined,
      "value": "accessToken"
    }, {
      "matches": 'https://api.redocly.com/registry/**',
      "name": "Authorization",
      "envVariable": undefined,
      "value": "accessToken"
    }]);
  });

  it('should resolve config http header by EU region', async () => {
    jest.spyOn(RedoclyClient.prototype, 'getTokens').mockImplementation(
      () => Promise.resolve([{ region: 'eu', token: "accessToken", valid: true }])
    );
    const config = await loadConfig();
    expect(config.resolve.http.headers).toStrictEqual([{
      "matches": 'https://api.eu.redocly.com/registry/**',
      "name": "Authorization",
      "envVariable": undefined,
      "value": "accessToken"
    }]);
  });
});
