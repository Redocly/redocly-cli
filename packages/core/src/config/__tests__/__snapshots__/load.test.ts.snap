// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createConfig > should create a config with the apis section 1`] = `
Config {
  "_alias": undefined,
  "_usedRules": Set {},
  "_usedVersions": Set {},
  "configPath": "redocly.yaml",
  "decorators": {
    "arazzo1": {},
    "async2": {},
    "async3": {},
    "oas2": {},
    "oas3_0": {},
    "oas3_1": {},
    "overlay1": {},
  },
  "doNotResolveExamples": false,
  "document": undefined,
  "extendPaths": [],
  "ignore": {},
  "pluginPaths": [],
  "plugins": [],
  "preprocessors": {
    "arazzo1": {},
    "async2": {},
    "async3": {},
    "oas2": {},
    "oas3_0": {},
    "oas3_1": {},
    "overlay1": {},
  },
  "rawConfig": {
    "apis": {
      "test@v1": {
        "root": "resources/pets.yaml",
        "rules": {
          "operation-summary": "warn",
          "rule/test": "warn",
        },
      },
    },
    "resolve": {
      "http": {
        "headers": [],
      },
    },
    "rules": {
      "no-empty-servers": "error",
      "operation-summary": "error",
      "rule/test": {
        "assertions": {
          "defined": true,
        },
        "subject": {
          "property": "x-test",
          "type": "Operation",
        },
      },
    },
    "telemetry": "on",
  },
  "resolve": {
    "http": {
      "customFetch": undefined,
      "headers": [],
    },
  },
  "resolvedConfig": {
    "apis": {
      "test@v1": {
        "arazzo1Decorators": {},
        "arazzo1Preprocessors": {},
        "arazzo1Rules": {},
        "async2Decorators": {},
        "async2Preprocessors": {},
        "async2Rules": {},
        "async3Decorators": {},
        "async3Preprocessors": {},
        "async3Rules": {},
        "decorators": {},
        "extendPaths": [],
        "oas2Decorators": {},
        "oas2Preprocessors": {},
        "oas2Rules": {},
        "oas3_0Decorators": {},
        "oas3_0Preprocessors": {},
        "oas3_0Rules": {},
        "oas3_1Decorators": {},
        "oas3_1Preprocessors": {},
        "oas3_1Rules": {},
        "overlay1Decorators": {},
        "overlay1Preprocessors": {},
        "overlay1Rules": {},
        "pluginPaths": [],
        "plugins": [],
        "preprocessors": {},
        "root": "resources/pets.yaml",
        "rules": {
          "assertions": [
            {
              "assertionId": "rule/test",
              "assertions": {
                "defined": true,
              },
              "severity": "warn",
              "subject": {
                "property": "x-test",
                "type": "Operation",
              },
            },
          ],
          "no-empty-servers": "error",
          "operation-summary": "warn",
        },
      },
    },
    "arazzo1Decorators": {},
    "arazzo1Preprocessors": {},
    "arazzo1Rules": {},
    "async2Decorators": {},
    "async2Preprocessors": {},
    "async2Rules": {},
    "async3Decorators": {},
    "async3Preprocessors": {},
    "async3Rules": {},
    "decorators": {},
    "extendPaths": [],
    "oas2Decorators": {},
    "oas2Preprocessors": {},
    "oas2Rules": {},
    "oas3_0Decorators": {},
    "oas3_0Preprocessors": {},
    "oas3_0Rules": {},
    "oas3_1Decorators": {},
    "oas3_1Preprocessors": {},
    "oas3_1Rules": {},
    "overlay1Decorators": {},
    "overlay1Preprocessors": {},
    "overlay1Rules": {},
    "pluginPaths": [],
    "plugins": [],
    "preprocessors": {},
    "resolve": {
      "http": {
        "headers": [],
      },
    },
    "rules": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "telemetry": "on",
  },
  "resolvedRefMap": undefined,
  "rules": {
    "arazzo1": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "async2": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "async3": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "oas2": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "oas3_0": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "oas3_1": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
    "overlay1": {
      "assertions": [
        {
          "assertionId": "rule/test",
          "assertions": {
            "defined": true,
          },
          "subject": {
            "property": "x-test",
            "type": "Operation",
          },
        },
      ],
      "no-empty-servers": "error",
      "operation-summary": "error",
    },
  },
}
`;
